app/code/Magento/GiftMessage/Test/Unit/Observer/SalesEventOrderItemToQuoteItemObserverTest.php

```diff
@@ -0,0 +1,196 @@
+use Magento\Quote\Model\Quote\Item as QuoteItem;
+use Magento\Sales\Model\Order;
+use Magento\Sales\Model\Order\Item as OrderItem;
+use Magento\Store\Model\Store;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ */
+class SalesEventOrderItemToQuoteItemObserverTest extends TestCase
+{
+    /**
+     * Stub message id
+     */
+    private const STUB_MESSAGE_ID = 1;
+
+    /**
+     * Stub new message id
+     */
+    private const STUB_NEW_MESSAGE_ID = 2;
+
+    /**
+     * @var SalesEventOrderItemToQuoteItemObserver
+     */
+    private $observer;
+
+    /**
+     * @var MessageFactory|MockObject
+     */
+    private $messageFactoryMock;
+
+    /**
+     * @var MessageHelper|MockObject
+     */
+    private $giftMessageHelperMock;
+
+    /**
+     * @var Observer|MockObject
+     */
+    private $observerMock;
+
+    /**
+     * @var Event|MockObject
+     */
+    private $eventMock;
+
+    /**
+     * @var Order|MockObject
+     */
+    private $orderMock;
+
+    /**
+     * @var OrderItem|MockObject
+     */
+    private $orderItemMock;
+
+    /**
+     * @var Store|MockObject
+     */
+    private $storeMock;
+
+    /**
+     * @var MessageInterface|MockObject
+     */
+    private $messageMock;
+
+    /**
+     * @var QuoteItem|MockObject
+     */
+    private $quoteItemMock;
+
+    /**
+     * Prepare environment for test
+     */
+    public function setUp(): void
+    {
+        $this->messageFactoryMock = $this->createMock(MessageFactory::class);
+        $this->giftMessageHelperMock = $this->createMock(MessageHelper::class);
+        $this->observerMock = $this->createMock(Observer::class);
+        $this->eventMock = $this->createPartialMock(Event::class, ['getOrderItem', 'getQuoteItem']);
+        $this->orderItemMock = $this->createPartialMock(
+            OrderItem::class,
+            ['getOrder', 'getStoreId', 'getGiftMessageId']
+        );
+        $this->quoteItemMock = $this->createPartialMock(QuoteItem::class, ['setGiftMessageId']);
+        $this->orderMock = $this->createPartialMock(Order::class, ['getReordered']);
+        $this->storeMock = $this->createMock(Store::class);
+        $this->messageMock = $this->createMock(MessageModel::class);
+
+        $objectManager = new ObjectManager($this);
+
+        $this->observer = $objectManager->getObject(
+            SalesEventOrderItemToQuoteItemObserver::class,
+            [
+                'messageFactory' => $this->messageFactoryMock,
+                'giftMessageMessage' => $this->giftMessageHelperMock
+            ]
+        );
+    }
+
+    /**
+     * Tests duplicating gift message from order item to quote item
+     *
+     * @param bool $orderIsReordered
+     * @param bool $isMessagesAllowed
+     * @dataProvider giftMessageDataProvider
+     */
+    public function testExecute($orderIsReordered, $isMessagesAllowed)
```

Comment:
1. Try to avoid having single method that tests all the cases.
2. Test method name should be explanatory what it checks.
3. Having single test method is very difficult to extend with other tests, because any try to add extra scenario requires from other person to refactor the mocks outside the test method.